### DISCUSSION TITLE
Can I iterate over all structs a user declared anywhere in his / her program?

### DISCUSSION THREAD

**Discussion by ghost** [odin-lang/Odin#980] [Q&A]
The idea is to iterate over all structs declared anywhere in the program, look for a specific tag, and do something based on that tag at runtime.

For example:

```Odin
Bar :: struct {
	z: int,
	w: string,
};

// ...

Foo :: struct `MyTag` {
	x: int,
	y: string,
	z: bool,
};

// ...

main :: proc() {
	for structure in AllStructsDeclaredAnywhereInThisProgram {
		structureTags := reflect.struct_tags(typeid_of(type_of(structure)));
		if "MyTag" in structureTags {
			doSomethingWithStructure(&structure);
		}
	}
}
```

Is this possible? If not, can it be theoretically possible at all? :)

Thanks!

**Comment by ghost** [comment]
I need `AllStructsDeclaredAnywhereInThisProgram` because a user can declare 128 526 structs in his / her 4 012 source file program, and I need to iterate over all of his / her structs to know which ones he / she tagged with `MyTag`, programmably.

**Comment by gingerBill** [comment]
@procedural Why not use the Odin parser and make a meta program for such a thing?

**Comment by gingerBill** [comment]
package odin/parser is a good place to start. You effectively parse the files you want and then generate the data for those things. i.e. make a mini compiler

**Comment by gingerBill** [comment]
@procedural You can add custom attributes to declarations and ignore them in normal compilation so you can analyse them in your meta program.
